libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/AppLockView/SFSDKBiometricViewController.m

```diff
@@ -0,0 +1,271 @@
+    CGFloat yCancelButton = CGRectGetMinY(self.setUpBiometricView.bounds) + hTitle + hIns + (kSFBioDefaultPadding * 2.5);
+    CGFloat wCancelButton = (self.view.bounds.size.width - (3 * kSFBioDefaultPadding)) / 2;
+    CGFloat hCancelButton = kSFBioButtonHeight;
+    self.cancelBiometricButton.frame = CGRectMake(xCancelButton, yCancelButton, wCancelButton, hCancelButton);
+    
+    CGFloat xEnableButton = CGRectGetMinX(self.setUpBiometricView.bounds) + (kSFBioDefaultPadding * 2) + wCancelButton;
+    CGFloat yEnableButton = CGRectGetMinY(self.setUpBiometricView.bounds) + hTitle + hIns + (kSFBioDefaultPadding * 2.5);
+    CGFloat wEnableButton = (self.view.bounds.size.width - (3 * kSFBioDefaultPadding)) / 2;
+    CGFloat hEnableButton = kSFBioButtonHeight;
+    self.enableBiometricButton.frame = CGRectMake(xEnableButton, yEnableButton, wEnableButton, hEnableButton);
+    
+    CGFloat xSetup = (0 - kSFViewBoarderWidth);
+    CGFloat ySetup = (kSFBioTopPadding * 2) + kSFIconCircleDiameter;
+    CGFloat wSetup = self.view.bounds.size.width + (kSFBioViewBorderWidth * 2);
+    CGFloat hSetup = (kSFBioDefaultPadding * 3.5) + hTitle + hIns + hCancelButton;
+    self.setUpBiometricView.frame = CGRectMake(xSetup, ySetup, wSetup, hSetup);
+}
+
+
+- (void)viewDidLoad {
+    self.view.backgroundColor = self.viewConfig.backgroundColor;
+    if ([self respondsToSelector:@selector(setEdgesForExtendedLayout:)]) {
+        [self setEdgesForExtendedLayout:UIRectEdgeNone];
+    }
+    [self layoutSetupBiometric];
+    self.navigationItem.hidesBackButton = YES;
+    
+    if (self.verificationMode) {
+         [self showBiometric];
+    } else {
+        [self showBiometricSetup];
+    }
+}
+
+// private methods
+- (void)showBiometric
+{
+    LAContext *context = [[LAContext alloc] init];
+    __weak typeof (self) weakSelf = self;
+    
+    if (self.verificationMode) {
+        [context setLocalizedCancelTitle:[SFSDKResourceUtils localizedString:@"biometricFallbackActionLabel"]];
+        [context setLocalizedFallbackTitle:@""];
+        [self hideAll];
+    }
+    
+    [context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics localizedReason:[SFSDKResourceUtils localizedString:@"biometricReason"] reply:^(BOOL success, NSError *authenticationError) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            __weak typeof (weakSelf) strongSelf = weakSelf;
+            if (success) {
+                [strongSelf.biometricResponseDelgate biometricUnlockSucceeded:strongSelf.currentPasscode  verificationMode:strongSelf.verificationMode];
+            } else {
+                [strongSelf.biometricResponseDelgate biometricUnlockFailed:strongSelf.currentPasscode verificationMode:strongSelf.verificationMode];
+            }
+        });
+    }];
+}
+
+- (void)userDenyBiometricEnablement
+{
+    [self.biometricResponseDelgate biometricUnlockFailed:self.currentPasscode verificationMode:self.verificationMode];
```

Comment:
It might be better to propagate the events that the biometricUnlockFailed  because the user Denied Enrollment or add another delegate method to allow  the  delegate to handle event for  "user denying biometric enrollment"